### ISSUE TITLE
bilateral-only "proof" of payment that's compatible w/ current protocol

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#2036]
Ticket #737 has a scope that allows changes to the protocol to implement a strong proof-of-payment protocol. This ticket is about a much more limited feature that can be implemented with the current protocol.

… needs specification

**Comment by zmanian** [comment]
What if instead of revealing ephemeral secret key, you did a XEDDSA signature with the Esk of (jsouput, recipient address, challenge)? https://whispersystems.org/docs/specifications/xeddsa/#xeddsa

**Comment by arielgabizon** [comment]
Maybe.. I'm slightly confused about what we need to defend against.
If it's a replay attack where some time later after I have sent my proof of payment, someone else replays that proof..can't the receiver just assume whomever sent the proof was the first time is the legitimate payment sender?
On the other hand, if it's a man-in-the-middle attack where someone is online between sender and receiver, he could still show that he payed even when there's an interactive signing challenge.
But I wonder, if for a first implementation, it's enough to just have what @nathan-at-least suggested above, and have the receiver assume whomever sent the proof of payment first is the legitimate one.

**Comment by zookozcash** [comment]
@arielgabizon: a good defense against those sorts of replay/MITM attacks is to include identifiers (e.g. public keys or addresses) of the prover and the verifier into the proof.

**Comment by arielgabizon** [comment]
That's a nice  idea. It can partially be implemented just with the memo field. Though a payer would have to know in advance he'll want to do such a proof of payment later so that he would include this receipt in the encrypted memo.

**Comment by daira** [comment]
@arielgabizon wrote:
> On the other hand, if it's a man-in-the-middle attack where someone is online between sender and receiver, he could still show that he payed even when there's an interactive signing challenge.

If you have a secure channel with a purported (potentially anonymous) sender, and they perform an interactive proof with you over that channel, then you know that subsequent communications on the channel are with the actual sender (or someone who knows the sender's secrets).

**Comment by jackgavigan** [comment]
@nathan-at-least wrote:
> A 'payment disclosure' includes: `(jsoutput, recipient address (pk_enc, a_pk), ephemeral secret key)`

Am I correct in thinking that the 'payment disclosure' consists of data that the Recipient will possess, so therefore a 'payment disclosure' can be created and shared by either the Sender or the Recipient?

**Comment by daira** [comment]
@jackgavigan The recipient doesn't know the ephemeral secret key (esk). It does know its own viewing key sk<sub>enc</sub> which would be sufficient to decrypt the ciphertext, but that is also sufficient to decrypt all other ciphertexts sent to the recipient address.

**Comment by bitcartel** [comment]
Could the following help increase the level of confidence that a presenter of the payment disclosure is indeed the sender (where the sender did not include anything in the memo field) ?

- Sender retains the ephemeral JoinSplitSig Private Key.
- Sender uses this private key to sign a payment disclosure / refund address / message.
- Recipient verifies signed data using the JoinSplitSig Public Key.

**Comment by bitcartel** [comment]
Draft ZIP avilable for review: https://github.com/zcash/zips/pull/119

**Comment by daira** [comment]
There's a problem I hadn't thought of (documented in the [draft zip here](https://github.com/bitcartel/zips/blob/9d0ff728d900d8dbb1c7e36cfcda72e411a98c80/drafts/bitcartel-payment-disclosure/draft.rst#known-issues)): in some cases (such as for t -> multiple z transactions), both outputs of each JoinSplit will be used as non-change outputs. In that case the "KDF Tweak" approach won't work, because that relies on the sender of a change output knowing sk<sub>enc</sub>.

There are several possible resolutions:
1. Make sure that each JoinSplit has at most one non-change output, at the expense of doubling the number of JoinSplits in the problematic case. E.g. for a transaction sending transparent input to two z-addresses, now two JoinSplits are needed instead of one.
2. Only reveal the symmetric key K<sup>enc</sup><sub>i</sub> (which is different for each output). This has the disadvantage that it is possible to provide a "valid" proof of payment, even though the recipient could not have decrypted the payment. The attacker would need to burn money in this case.
3. A payment disclosure contains the inputs to the KDF, i.e. (i, hSig, sharedSecret<sub>i</sub>, epk, pk<sup>new</sup><sub>enc,i</sub>), and a proof (e.g. Chaum-Pedersen) that (g, epk, pk<sup>new</sup><sub>enc,i</sub>, sharedSecret<sub>i</sub>) is a DH tuple. Note that to avoid an information leak, we need that pk<sup>new</sup><sub>enc,0</sub> and pk<sup>new</sup><sub>enc,1</sub> are distinct (because if they are equal, then sharedSecret<sub>0<sub> = sharedSecret<sub>1</sub>). But it is easy to avoid this case.

**Comment by bitcartel** [comment]
In #2 an independent judge/arbiter would require the recipient to provide a viewing key, if the recipient continues to claim they did not receive funds.  So there could be a situation where an attacker burns funds, opens a dispute, provides an invalid proof of payment, with the intent to get the recipient to reveal their viewing key.

**Comment by daira** [comment]
Actually in case 2., revealing K<sup>enc</sup><sub>i</sub> would then allow the recipient to spend the funds. So the issue is that the recipient would not previously have been able to spend them, and the payment might have been time-sensitive.

**Comment by daira** [comment]
Here's the original paper on Chaum-Pedersen: http://www.cs.elte.hu/~rfid/chaum_pedersen.pdf

It's presented as a signature scheme, but one that is constructed from a proof-of-DH-tuple. There are other possible ways to prove a DH tuple based on pairings, but Curve25519 isn't pairing-friendly.

**Comment by daira** [comment]
A possible refinement of 3. is to use a Chaum-Pedersen signature to sign the information that should be nonmalleable (e.g. refund address). This is similar to @zmanian's idea of using XEdDSA.

**Comment by zmanian** [comment]
Mike Hamburg has an x25519 signatures scheme in Strobe.
https://www.reddit.com/r/crypto/comments/5u7i3z/signature_with_a_montgomery_curve/ddu8or0/?context=3 See @bascule's comment and Mike's response

**Comment by daira** [comment]
In https://github.com/zcash/zcash/issues/1360#issuecomment-282460362 I point out that using both outputs of a JoinSplit entails an information leak. So there is a privacy-performance trade-off which, if we resolved it in favour of privacy at the cost of sometimes using more JoinSplits, would also allow us to use the simpler KDF Tweak approach (#2102) for payment disclosures.

**Comment by daira** [comment]
In the next circuit change, we can eliminate this privacy-performance tension by implementing #647, as I pointed out in https://github.com/zcash/zcash/issues/647#issuecomment-246174077 .

**Comment by daira** [comment]
I think we should exclude using an unconventional signature scheme. We have the joinSplitSig private Ed25519 key available to sign the associated nonmalleable information. Ed25519 is pleasantly boring.

**Comment by arielgabizon** [comment]
The problem with option two may be smaller than we originally thought.
As I think the payment disclosure K^i_enc give the recipient the ability to decode and spend the note, even if they couldn't initially.
So if a third party receives the payment disclosure, they could send it to the recipient to ensure the recipient can spend the note, in case they are concerned about an attack as explained in above.

**Comment by daira** [comment]
Yep, that's what I said in https://github.com/zcash/zcash/issues/2036#issuecomment-282360893

**Comment by arielgabizon** [comment]
Right, maybe can just delete these last 3 comments

**Comment by nathan-at-least** [comment]
Let's implement the new multi-joinsplit algorithm as an opt-in 'experimental feature', and point out it has better privacy and worse performance in the specific cases of many recipients (and many input notes). Then we'll get feedback from the wild-and-wooly about this.

**Comment by daira** [comment]
The worse performance is only when there are many output notes (and fewer than twice as many input notes than output notes).

**Comment by arielgabizon** [comment]
@nathan-at-least  - Not sure why you posted this here. Does that mean we implement payment disclosure only after implementing new multi-joinsplit?
Can someone link to a description of this new algorithm?

**Comment by daira** [comment]
@arielgabizon wrote:
> Can someone link here to a description of this new multi join-split algorithm?

https://github.com/zcash/zcash/issues/1360#issuecomment-274644332

The relevance to payment disclosure is that if we didn't the new multijoinsplit algorithm, then a payment disclosure would leak the other JoinSplit output (the one not intended to be disclosed)

**Comment by nathan-at-least** [comment]
Let's release this without fixing the privacy leaks as an experimental feature (with very clear caution text about privacy leaks) in 1.0.8, then fix privacy leaks as soon as we can after that.

**Comment by arielgabizon** [comment]
> The relevance to payment disclosure is that if we didn't implement the new multijoinsplit algorithm (and also #2102), then a payment disclosure would leak the other JoinSplit output — the one not intended to be disclosed.
@daira there is no leak in option 3 above, but I guess we've decided it's too complicated for now to add this nizk?

**Comment by daira** [comment]
I think we decided that we were going to implement the new multijoinsplit algorithm in such a way that no more than one output of each JoinSplit would need a payment disclosure (because that reduces information leakage in v<sup>pub</sup><sub>in,out</sub> values), and therefore that option 3 is not needed.

**Comment by arielgabizon** [comment]
OK, I guess if we're implementing JoinSplit that way it will be fine to give e<sub>sk</sub>

**Comment by daira** [comment]
@ebfull wrote:
> Could the following help increase the level of confidence that a presenter of the payment disclosure is indeed the sender (where the sender did not include anything in the memo field) ?
>
> * Sender retains the ephemeral JoinSplitSig Private Key.
> * Sender uses this private key to sign a payment disclosure / refund address / message.
> * Recipient verifies signed data using the JoinSplitSig Public Key.

Yes, that's what I suggested in https://github.com/zcash/zcash/issues/737#issuecomment-192359314 . @nathan-at-least opened this ticket to be for something simpler. However I think we'll end up adding a signature, just not in the initial implementation.

**Comment by bitcartel** [comment]
@daira et al remind me.. why can't we simply retain a record of the symmetric key used for each call to `crypto_aead_chacha20poly1305_ietf_encrypt`?  In `JoinSplit.cpp` we use the same encryptor object for both note outputs, where for each output, the nonce input to the KDF is incremented.  The resulting symmetric key is different for each note output.

**Comment by daira** [comment]
Only recording the secret key would not be sufficient information to prove that the recipient could have decrypted the note ciphertext from the block chain.

I was thinking that the payment disclosure would include pk<sub>enc</sub>, because it's supposed to be a proof that a given address received a payment, which necessarily reveals the address.

**Comment by bitcartel** [comment]
Wouldn't an arbitrator need the full payment address, rather than just pk<sub>enc</sub> ?

**Comment by daira** [comment]
Yes, but given esk and pk<sub>enc</sub> they can decrypt that, they don't need to be given it explicitly. (If they were given it then they would still have to check it against the decrypted version.)

**Comment by nathan-at-least** [comment]
Bumped out of 1.0.9 which is focused on stabilization rather than big new features.

**Comment by bitcartel** [comment]
> doesn't include whether it is a mainnet or testnet z-address. So the payment disclosure does need to include such information.

Should we offer two types of payment blob?

One which provides complete context, useful for a third party, by including apk so that the third party knows if this is mainnet or testnet.

Another which provides less context, by not including apk, so that two parties who want to confirm payment between themselves but do not want to reveal the z address to an attacker who intercepts the payment blob?

**Comment by bitcartel** [comment]
Including the apk does make the user interface easier on the validation side, as you don't have to include it.

**Comment by bitcartel** [comment]
@daira So to clarify, the payment disclosure should include the recipient's shielded payment address. 
 This would allow the recipient to verify that the payment disclosure belongs to them (since they have the spending key in their wallet for that payment address).  It also means that any person looking to validate the payment disclosure would have a<sub>pk</sub> and pk<sub>enc</sub>, where pk<sub>enc</sub> is required for decrypting the ciphertext.

**Comment by tromer** [comment]
Another attack on the ZIP's protocol, again exploiting inconsistency between ciphertext and JoinSplit:

The sender broadcasts a JoinSplit creating a low-value note, but attaches a ciphertext which is the encryption of a note plaintext claiming high value. He can then claim that he sent a high-value note, and present a payment disclosure that passes verification. But the recipient never received a high-value note, and the sender never nullified one.

**Comment by daira** [comment]
Yes, I believe that payment disclosure proofs that are robust against collusion between the sender and recipient will only be possible with Sapling.

(In the current proposal on #2277, the correctness of epk and the note commitment, which contains esk, are both checked by the circuit, and so even though the correctness of the symmetric encryption isn't directly checked, it must be the case that either the note ciphertext is correct or the payment proof verifier can detect that it is not.)

**Comment by tromer** [comment]
@daira, the attack in https://github.com/zcash/zcash/issues/2036#issuecomment-329872792 doesn't require collusion.

In any case, the ZIP must be clearer about the claimed security properties of the protocol.

**Comment by daira** [comment]
I don't think that attack works: the note commitment revealed by the JoinSplit won't match the one computed from the decrypted note plaintext.

**Comment by tromer** [comment]
The ZIP is ambiguous on what check is done (see https://github.com/zcash/zips/pull/119/files#r139325414).

**Comment by tromer** [comment]
Can the *receiver* of a transaction generate a payment disclosure, that will convince a third party?
The ZIP (https://github.com/zcash/zips/pull/119/files) and doc (https://github.com/zcash/zcash/blob/master/doc/payment-disclosure.md) are both silent/ambiguous about this point.

**Comment by daira** [comment]
No, because they don't have the JoinSplit signature private key. I agree that the ZIP should say this explicitly. Note that anyone can *replay* a payment disclosure, but only with a message that was signed by a holder of the JoinSplit signature private key. So it is possible to do an interactive proof of payment by requiring a fresh nonce to be included in the message.

**Comment by tromer** [comment]
This issue is out of date. Sapling changed the underlying protocol, and does not specify any form of payment disclosures yet.

Meanwhile, another feature to consider is: can payment disclosures be used to detect when the note is _spent_? The spec should say explicitly say either that _yes_ (and specify how, and guarantee completeness and soundness), or _no_ (and guarantee privacy).

**Comment by tromer** [comment]
Two variants are sketched by @str4d [here](https://forum.zcashcommunity.com/t/z-addresses-usage-analysis/35497/9).

Everything said in #292 applies here too: we need to carefully define, prove and communicate the security properties of the chosen protocol.

**Comment by daira** [comment]
I'm marking this as blocking https://github.com/zcash/zips/issues/387 because we should go through the discussion and decide whether any of it is relevant to ZIP 311.
